lib/rules/no-undefined.js

```diff
@@ -21,15 +21,38 @@ module.exports = {
 
     create(context) {
 
+        /**
+         * Returns true if an Identifier with name of "undefined" is in an
+         * acceptable location.
+         * @param {ASTNode} node The Identifer node to check.
+         * @returns {boolean} True if undefined is allowed here, false otherwise.
+         */
+        function isValidUndefinedIdentifier(node) {
+            switch (node.parent.type) {
+                case "MemberExpression":
+
+                    // foo.undefined is valid, foo[undefined] is not
+                    return node === node.parent.property && !node.parent.computed;
+
+                case "Property":
+
+                    // Only non-computed keys are valid here
+                    return node === node.parent.key &&
+                        !node.parent.computed;
```

Comment:
I think this needs to include a `&& !node.parent.shorthand` check, otherwise the rule will have a false negative with

```js
({ undefined });

// this should get reported, because it's equivalent to

({ undefined: undefined });
```